resource-providers/src/main/java/io/opentelemetry/resource/spring/SpringBootServiceNameGuesser.java

```diff
@@ -0,0 +1,255 @@
+    } catch (Exception e) {
+      // expected to fail sometimes
+    }
+    logger.debug("Checking application.yml in current dir: " + result);
+    return result;
+  }
+
+  @Nullable
+  @SuppressWarnings("unchecked")
+  private static String parseNameFromYaml(InputStream in) {
+    Yaml yaml = new Yaml();
+    try {
+      Map<String, Object> data = yaml.load(in);
+      Map<String, Map<String, Object>> spring =
+          (Map<String, Map<String, Object>>) data.get("spring");
+      if (spring != null) {
+        Map<String, Object> app = spring.get("application");
+        if (app != null) {
+          Object name = app.get("name");
+          if (name != null) {
+            return (String) name;
+          }
+        }
+      }
+    } catch (RuntimeException e) {
+      // expected to fail sometimes
+    }
+    return null;
+  }
+
+  @Nullable
+  private String findByCommandlineArgument() {
+    String result = attemptProcessHandleReflection();
+    if (result == null) {
+      String javaCommand = system.getProperty("sun.java.command");
+      result = parseNameFromCommandLine(javaCommand);
+    }
+    logger.debug("Checking application commandline args: " + result);
+    return result;
+  }
+
+  @Nullable
+  private String attemptProcessHandleReflection() {
+    try {
+      String commandLine = system.attemptGetCommandLineViaReflection();
+      return parseNameFromCommandLine(commandLine);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  private static String parseNameFromCommandLine(String commandLine) {
+    Matcher matcher = COMMANDLINE_PATTERN.matcher(commandLine);
+    boolean ignored = matcher.find(); // Required before group()
+    return matcher.group(1);
+  }
+
+  @Nullable
+  private String readNameFromAppProperties() {
+    return loadFromClasspath(
+        "application.properties", SpringBootServiceNameGuesser::getAppNamePropertyFromStream);
+  }
+
+  @Nullable
+  private static String getAppNamePropertyFromStream(InputStream in) {
+    Properties properties = new Properties();
+    try {
+      properties.load(in);
+      return properties.getProperty("spring.application.name");
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  @Nullable
+  private String loadFromClasspath(String filename, Function<InputStream, String> parser) {
+    try (InputStream in = system.openClasspathResource(filename)) {
+      return parser.apply(in);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  // Exists for testing
+  static class SystemHelper {
+
+    String getenv(String name) {
+      return System.getenv(name);
+    }
+
+    String getProperty(String key) {
+      return System.getProperty(key);
+    }
+
+    InputStream openClasspathResource(String filename) {
+      return getClass().getClassLoader().getResourceAsStream(filename);
+    }
+
+    InputStream openFile(String filename) throws Exception {
+      return Files.newInputStream(Paths.get(filename));
```

Comment:
I would check that the file exists before attempting to open it

Reply:
Why, just to avoid the FileNotFound exception? I'm pretty "meh" on this. 

I mean, if we're going to do that we might as well check to ensure that the file exists AND that it's a regular file AND that I have permission to read it also, right?

Reply:
This can be super annoying if you need to debug some exception during agent startup and use an exception breakpoint and you get a million random exceptions before the one where you want to stop. Imo it is best not to cause exceptions if you can easily avoid it. I'd add regular file check, but I wouldn't bother with checking permissions.

Reply:
Is it more annoying that code that's just littered with checked exception handlers and guarded code all over the place? 😁 

In this case, if I add check, what should I do in the event that the file doesn't exist? 

Apparently I shouldn't throw a different exception, so I guess I slap `@Nullable` on the method and return a `null` `InputStream`? That will cause an NPE in the caller unless guarded, and the callers are currently using try-with-resources. So I'll have to unwrap those into separate assignments. Yuck. Alternatively, I could return a `ByteArrayInputStream` with nothing in it and hope that the parsers can grok an empty stream without exploding -- but returning an empty stream also suggests to any caller that the file was actually read but contains nothing. That's not great either. I dunno. Situations like this always feel like a language shortcoming to me.

Coincidentally, I was reading the [nodejs docs on `fs.exists()`](https://nodejs.org/api/fs.html#fsexistspath-callback) yesterday and they explicitly discourage checking before attempting because it's not multi-thread/process safe. That doesn't really apply here, but I thought it was interesting nonetheless.

Reply:
As far as I know try with resource works nicely with `null`, it won't NPE.